### PR TITLE
Rule S2187: Fix FP, check all upper class hierarchy for test attributes

### LINKED ISSUE #5507: S2187: Rule should not raise False Positives if MSTEST test method is in a base class higher up in the inheritance hierarchy
### Description

I am getting S2187 if a test method is provided in an abstract class higher up the inheritance hierarchy only.

### Repro steps

```
	[TestClass]
	public abstract class A
	{
		[TestMethod]
		public void Test()
		{
		}
	}

	[TestClass]
	public abstract class B : A
	{
		// works if this is uncommented
		//[TestMethod]
		//public void Test()
		//{
		//}
	}

	[TestClass]
	public class C : B
	{

	}
```

### Expected behavior

No warning.

### Actual behavior

Warning about S2187 on class C.

### Known workarounds

Suppress warning

### Related information

SonarLint 5.5.0.43817
SonarQube 8.9.6.50800
Visual Studio 17.1.2

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes #5507

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/TestClassShouldHaveTestMethod.cs [hunk 1]
```diff
@@ -65,19 +65,34 @@ public sealed class TestClassShouldHaveTestMethod : SonarDiagnosticAnalyzer
...
+            symbol.IsTestClass()
+            && !HasAnyTestMethod(symbol);
 
-        private static bool IsExceptionToTheRule(INamedTypeSymbol classSymbol) =>
-            classSymbol.IsAbstract
-            || (classSymbol.BaseType.IsAbstract && HasAnyTestMethod(classSymbol.BaseType))
-            || HasSetupOrCleanupAttributes(classSymbol);
+        private static bool IsExceptionToTheRule(ITypeSymbol symbol) =>
+            symbol.IsAbstract
+            || HasBaseTypeWithTestMethods(symbol)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/TestClassShouldHaveTestMethod.cs:77]: We have `symbol.GetSelfAndBaseTypes()` that we can use. So this could look something like
```
|| symbol.GetSelfAndBaseTypes().Any(HasAnyTestMethod)
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
We can delete some code 😄

File: analyzers/src/SonarAnalyzer.CSharp/Rules/TestClassShouldHaveTestMethod.cs [hunk 2]
```diff
@@ -65,19 +65,19 @@ public sealed class TestClassShouldHaveTestMethod : SonarDiagnosticAnalyzer
                 SyntaxKind.ClassDeclaration,
                 SyntaxKindEx.RecordClassDeclaration);
 
-        private static bool HasAnyTestMethod(INamedTypeSymbol classSymbol) =>
-            classSymbol.GetMembers().OfType<IMethodSymbol>().Any(m => m.IsTestMethod());
+        private static bool HasAnyTestMethod(INamespaceOrTypeSymbol symbol) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/TestClassShouldHaveTestMethod.cs:68]: Why this change from `INamedTypeSymbol` to `INamspaceOrTypeSymbol`? It is not used inside, but from a method contract point of view, this method doesn't make sense for namespaces.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.CSharp/Rules/TestClassShouldHaveTestMethod.cs (hunk 2 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/TestClassShouldHaveTestMethod.cs:68]: I am using the most abstract type for flexibility. No strong preference here.
